src/Analysis/Ast/Impl/Analyzer/Caching/CacheFolders.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.IO;
+using Microsoft.Python.Core.Logging;
+using Microsoft.Python.Core.OS;
+
+namespace Microsoft.Python.Analysis.Analyzer.Caching {
+    internal static class CacheFolders {
+        public static string GetCacheFilePath(string root, string moduleName, string content, IFileSystem fs) {
+            // Folder for all module versions and variants
+            // {root}/module_name/content_hash.pyi
+            var folder = Path.Combine(root, moduleName);
+
+            var filePath = Path.Combine(root, folder, $"{FileNameFromContent(content)}.pyi");
+            if (fs.StringComparison == StringComparison.Ordinal) {
+                filePath = filePath.ToLowerInvariant();
+            }
+            return filePath;
+        }
+
+        public static string GetCacheFolder(IServiceContainer services) {
+            var platform = services.GetService<IOSPlatform>();
+            var logger = services.GetService<ILogger>();
+            
+            // Default. Not ideal on all platforms, but used as a fall back.
+            var localAppData = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+            var defaultCachePath = Path.Combine(localAppData, "Microsoft", "Python Language Server");
+
+            string path = null;
+            try {
+                const string plsSubfolder = "Microsoft/Python.Language.Server";
+                if (platform.IsMac) {
+                    var expanded = Environment.ExpandEnvironmentVariables($"$(HOME)/Library/Caches/{plsSubfolder}");
```

Comment:
`ExpandEnvironmentVariables` assumes that the input is in `%FOO_BAR%` form. On my Linux machine:

```
'$XDG_CACHE_HOME' expands to '$XDG_CACHE_HOME'
'$HOME' expands to '$HOME'
'%XDG_CACHE_HOME%' expands to '%XDG_CACHE_HOME%'
'%HOME%/.cache' expands to '/home/jake/.cache'
```

Also, `ExpandEnvironmentVariables` doesn't return null if something isn't found; it just doesn't replace it (hence why that third example breaks on my system where that variable is not set). I think you're going to have to use `GetEnvironmentVariable` and join yourself to ensure that the variables exist before choosing a path.

Reply:
Oh, and when you check `GetEnvironmentVariable`, I'd check `string.IsNullOrEmpty` instead of just null (since a variable can be set but empty).

Reply:
https://github.com/dotnet/corefx/issues/28890 suggests `Replace('$', '%')`

Reply:
Yeah, I wouldn't do that, since there would be no terminating `%`. With the change you just made it'll work so long as the variables exist, but because `ExpandEnvironmentVariables` will just not replace things if they aren't set I think you're still going to have to look up variables first and then Combine them.